README
Rest-HtmlVis

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

BUGS

Please report any bugs or feature requests to github repository.

REPOSITORY

L<https://github.com/vasekd/Rest-HtmlVis>

LICENSE AND COPYRIGHT

Copyright 2015 Vaclav Dovrtel.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.



Changes
Revision history for Rest-HtmlVis

{{$NEXT}}

0.13  2016-04-14 16:45:05 CEST
	Fixed missing Footer

0.12  2016-04-11 15:32:10 CEST
	Added footer support
	Fixed default params in form

0.11  2016-04-09 18:51:09 CEST
	Added support fir GET_FORM
	Added error handling

0.10  2016-04-09 00:45:39 CEST
	Added support for utf8
	Rewrite Content block
	Set right urlmap

0.09  2016-04-08 02:46:14 CEST
	Added yaml parser for struct

0.08  2015-09-08 14:51:57 CEST
	Form configuration is case insensitive
	Added "Get As" to post form

0.07  2015-09-03 13:54:05 CEST
	Fixed baseurl link in conf
	Changed links to link

0.06  2015-08-28 16:20:13 CEST
	Added radio and checkbox form elements

0.05  2015-05-20 18:52:42 CEST
	Added env to GET_FORM
	Added select element for GET_FORM
	Fixed empty content

0.04  2015-04-02 10:17:30 CEST
	Fixed global variable

0.03  2015-03-31 19:10:43 CEST
	Changed default static path to htmlvis, for better compatibility
	Added missing bootstrap directory

0.02  2015-03-24 02:29:48 CET
	Switch to milla

0.01    4. Mar 2015
        First version.



lib/Rest/HtmlVis.pm
package Rest::HtmlVis;

use 5.008_005;
use strict;
use warnings FATAL => 'all';

our $VERSION = '0.12'; # Set automatically by milla

my $based = {
	'default.base' => 'Rest::HtmlVis::Base',
	'default.content' => 'Rest::HtmlVis::Content',
	'default.footer' => 'Rest::HtmlVis::Footer',
};

sub new {
	my ($class, $params) = @_;

	my $htmlVis;

	my $self = bless {}, $class;

	### Set uri path
	$self->{baseurl} = "/static";
	# Don't delete key, because of original hash
	$self->{baseurl} = $params->{'default.baseurl'} if exists $params->{'default.baseurl'};

	### Add htmlvis
	foreach my $key (sort keys %$params){
		next if $key eq 'default.baseurl';
		$self->loadVisObject($key, $params->{$key});
	}

	### Set params
	foreach my $key (sort keys %$based) {
		$self->loadVisObject($key, $based->{$key}) unless exists $params->{$key};
	}

	return $self;
}

sub baseurl {
	my ($self) = shift;
	return $self->{baseurl};
}

sub loadVisObject {
	my ($self, $key, $class) = @_;

	my ($rtrn, $err) = _try_load($class);
	if ($rtrn){
		my $vis = $class->new($self->baseurl);
		my $order = $vis->getOrder;
		push(@{$self->{htmlVis}{$order}}, {
			key => $key,
			object => $vis
		}) if $vis->isa('Rest::HtmlVis::Key');
	}else{
		print STDERR "ERROR: to load vis class: ".$err."\n";
	}
}

sub html {
	my ($self, $struct, $env, $header) = @_;

	### manage keys
	my $head_parts = '';
	my $onload_parts = '';
	my $html_parts = '';
	my $footer_parts = '';

	my $rowBlocks = 0; # count number of blocks in row

	### Add blocks
	foreach my $order (sort keys %{$self->{htmlVis}}) {
		foreach my $obj (@{$self->{htmlVis}{$order}}) {

			my $vis = $obj->{object};
			next unless $vis->setStruct($obj->{key}, $struct, $env);
			$vis->setHeader($header);

			my $head = $vis->head($self->{local});
			$head_parts .= $head if $head;

			my $onload = $vis->onload();
			$onload_parts .= $onload if $onload;

			my $html = $vis->html();
			if ($html){
				$rowBlocks += $vis->blocks();
				my $newRow = ($vis->newRow() or $rowBlocks > 12) ? 1 : 0;

				$html_parts .= '<div class="row">' if $newRow;
				$html_parts .= $html;
				$html_parts .= '</div>' if $newRow;
				$rowBlocks = 0 if $newRow;
			}
			my $footer = $vis->footer($self->{local});
			$head_parts .= $footer if $footer;
		}
	}

	return "<!DOCTYPE html>\n<html>\n<head>\n$head_parts\n</head>\n<body onload=\"$onload_parts\">\n$html_parts\n</body>\n</html>";
}

### Try load library
sub _try_load {
	my $mod = shift;

	return +(0, "Not defined module.") unless $mod;
	return 1 if ($mod->can("html")); # because of local class in psgi
	eval("use $mod; 1") ? return 1 : return (0, $@);
}

1;

=encoding utf-8

=head1 NAME

Rest::HtmlVis - Rest API visualizer in HTML

=head1 SYNOPSIS

Transform perl hash to html.
Each key in perl hash is transormed to the piece of html, js and css which are include in main html.

Example:

	use Rest::HtmlVis;

	my $htmlvis = Rest::HtmlVis->new({
		events => Rest::HtmlVis::Events
	});

	$htmlvis->html({

		events => [
		],

		links => {
			rel => 'root',
			href => /,
			name => Root resource
		}

		form => {
			GET => {
				from => {
					type => 'time',
					default => time(),
				}
			},
			POST => {
				DATA => {
					type => "text"
					temperature => 25
				},

			}
		}
	});


HtmlVis has default blocks that are show everytime:

=over 4

=item * default.baseurl

Set default prefix for links in html. Default is '/static'

=item * default.base

See L<Rest::HtmlVis::Base>.

=item * default.content

See L<Rest::HtmlVis::Content>.

=back

These blocks can be rewrite when the base or content key is set in constructor params.

=head1 SUBROUTINES/METHODS

=head2 new( params )

Create new htmlvis object. You have to specify params for keys that should be transformed.

=head3 params

Define keys in input hash and library that manage this key.

Example:

	{ events => Rest::HtmlVis::Events }   

Third-party js library are primary mapped to /static URL.
You have to manage this url by your http server and map this url to share directory.
For example in Plack:
	
	use File::ShareDir;
	my $share = File::ShareDir::dist_dir('Rest-HtmlVis') || "../Rest-HtmlVis/share/";
	mount "/static" => Plack::App::File->new(root => $share);

=cut

=head2 html( hash_struct )

Convert input hash struct to html. Return html string.

=cut

=head1 TUTORIAL

L<http://psgirestapi.dovrtel.cz/>

=head1 AUTHOR

Václav Dovrtěl E<lt>vaclav.dovrtel@gmail.comE<gt>

=head1 BUGS

Please report any bugs or feature requests to github repository.

=head1 ACKNOWLEDGEMENTS

Inspired by L<https://github.com/towhans/hochschober>

=head1 COPYRIGHT

Copyright 2015- Václav Dovrtěl

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Rest/HtmlVis/Key.pm
package Rest::HtmlVis::Key;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

Rest::HtmlVis::Key - Base class for easy-to-use html vis

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.12';


=head1 SYNOPSIS

All you have to do is to inherit from Rest::HtmlVis::Key and then implement the callback html.

Example:

    package Rest::HtmlVis::MyGraph;
	use parent qw( Rest::HtmlVis::Key );
	
	use YAML::Syck;

	sub html {
		my ($self) = @_;

		local $Data::Dumper::Indent=1;
		local $Data::Dumper::Quotekeys=0;
		local $Data::Dumper::Terse=1;
		local $Data::Dumper::Sortkeys=1;

		return '<div class="col-lg-12">'.Dump($self->getStruct).'</div>';
	}

=head1 SUBROUTINES/METHODS

=head2 new

=cut

sub new {
	my ($class, $baseurl) = @_;
	return bless {baseurl => $baseurl}, $class;
}

=head2 baseurl

=cut

sub baseurl {
	return $_[0]->{baseurl};
}

=head2 setHeader

=cut

sub setHeader{
	my ($self, $header) = @_;
	my %hash = @{$header};
	$self->{header} = \%hash;
}

=head2 getHeader

=cut

sub getHeader {
	my ($self) = @_;
	return $self->{header};
}

=head2 setStruct

=cut

sub setStruct {
	my ($self, $key, $struct, $env) = @_;
	if ($struct && ref $struct eq 'HASH' && exists $struct->{$key}){
		$self->{struct} = $struct->{$key};
		$self->{env} = $env;
		return 1;
	}else{
		$self->{struct} = undef;
	}
	return undef;
}

=head2 getStruct

Return html hash with input structure.

=cut

sub getStruct {
	my ($self) = @_;
	return $self->{struct};
}

=head2 getEnv

Return env variables.

=cut

sub getEnv {
	my ($self) = @_;
	return $self->{env};
}

=head2 getOrder

Return wight on elemnt on html page. Default 1 means in middle;

=cut

sub getOrder {
	return 1;
}

=head2 blocks

Number of blocks in row. Max 12. (ala bootstrap)

Default is 12;

=cut

sub blocks {
	my ($self) = @_;
	return 12;
}

=head2 newRow

Define if element is on new row or is the part of previous row.

Default is 0 - no new row;

=cut

sub newRow {
	my ($self) = @_;
	return 0;
}

=head2 head

Return head of page as HTML string.

Default empty.

=cut

sub head {
	my ($self) = @_;
	return;
}

=head2 head

Return footer of page as HTML string.

Default empty.

=cut

sub footer {
	my ($self) = @_;
	return;
}

=head2 onload

Return onload javascript function of onload attr in body. It must ends with ;

Default empty.

=cut

sub onload {
	my ($self) = @_;
	return;
}

=head2 html

Return body part of HTML.

Default empty.

=cut

sub html {
	my ($self) = @_;
	return;
}

=encoding utf-8

=head1 AUTHOR

Václav Dovrtěl E<lt>vaclav.dovrtel@gmail.comE<gt>

=head1 BUGS

Please report any bugs or feature requests to github repository.

=head1 ACKNOWLEDGEMENTS

Inspired by L<https://github.com/towhans/hochschober>

=head1 REPOSITORY

L<https://github.com/vasekd/Rest-HtmlVis>

=head1 LICENSE AND COPYRIGHT

Copyright 2015 Vaclav Dovrtel.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Rest::HtmlVis::Key


lib/Rest/HtmlVis/Base.pm
package Rest::HtmlVis::Base;

use 5.006;
use strict;
use warnings FATAL => 'all';

use parent qw( Rest::HtmlVis::Key );

=head1 NAME

Rest::HtmlVis::Base - Return base struct of rest html (menu, styles etc.)

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.12';

sub setStruct {
	my ($self, $key, $struct, $env) = @_;
	$self->{struct} = $struct;
	$self->{env} = $env;
	return 1;
}

sub getOrder {
	return 0;
}

sub blocks {
	return 0;
}

my $style = 
'		body {
		  padding: 0px;
		  margin: 0px;
		}

		/* bootstrap-3.3.2-dist */
		.row {
		  margin: 10px 0px 0px 0px;
		}

		.form-control {
			margin-bottom: 4px;
		}

		/* Header styles */
		.header {
		  background-color: #eee;
		  display: inline-block;
		  width: 100%;
		  min-height: 38px;
		  padding: 6px 0px 5px 0px;
		  border: 1px solid #ddd
		}

		/* Header title style */
		.header .title {
		  margin: 9px;
		  display:inline;  
		  text-decoration: none;
		  color: #aaa;
		  font-size: 24px;
		}

		.title .project {
		  color: #ccc;
		}

		/* Format styles */
		.format {
		  float:right;
		  list-style-type: none;
		  margin: 5px 10px 0px 0px;
		  padding: 0px;
		}

		.format li {
		  float:left;
		  border-right: 1px solid #fff;
		  padding: 0px 7px 0px 7px;
		}

		.format li:last-child {
		  border-right: none;
		}

		.format li a {
		  margin-top: 1px;
		  text-decoration: none;
		  color: #ccc;
		  text-transform: uppercase;
		 
		  -webkit-transition: all 0.5s ease;
		  -moz-transition: all 0.5s ease;
		  -o-transition: all 0.5s ease;
		  -ms-transition: all 0.5s ease;
		  transition: all 0.5s ease;
		}
		 
		.format li a:hover {
		  color: #666;
		}
		 
		.format li.active a {
		  font-weight: bold;
		  color: #333;
		}

		/* Links ul */
		.links {
		  list-style-type: none;
		  margin: 5px 0px 0px 10px;
		  padding: 0px;
		  font: 14px/1 Helvetica, Verdana, sans-serif;
		}

		.links li {
		  padding-bottom: 1px;
		}

		.links li a {
		  margin-top: 1px;
		  text-decoration: none;
		  color: #23527c;
		 
		  -webkit-transition: all 0.5s ease;
		  -moz-transition: all 0.5s ease;
		  -o-transition: all 0.5s ease;
		  -ms-transition: all 0.5s ease;
		  transition: all 0.5s ease;
		}
		 
		.links li a:hover {
		  color: #666;
		}
		 
		.links li.active a {
		  font-weight: bold;
		  color: #333;
		}

		.method-form {
		  margin: 7px 0px 5px 0px;
		}

		.links span {
		  color: #aaa;
		}

		/* pretty print */
		pre.prettyprint {
		  padding: 10px;
		  border: 1px solid #ccc;
		}

		.kwd {
		  color: #23527c;
		}

		a {
		  color: #23527c;
		}
';

sub head {
	my ($self, $local) = @_;

	my $static = $self->baseurl;

	return '
	<meta charset="UTF-8">
	<link href="'.$static.'/google-code-prettify/prettify.css" rel="stylesheet">
	<link href="'.$static.'/bootstrap-3.3.2-dist/css/bootstrap.min.css" rel="stylesheet">
	<style>
	'.$style.'
	</style>	

	<meta name="viewport" content="width=device-width, initial-scale=1">

	<script src="'.$static.'/jquery.min.js"></script>
	<script src="'.$static.'/bootstrap-3.3.2-dist/js/bootstrap.min.js"></script>
	<script src="'.$static.'/google-code-prettify/prettify.js"></script>
	<script src="'.$static.'/google-code-prettify/lang-yaml.js"></script>
		';

}

sub html {
	my ($self) = @_;

	my $project = $self->getEnv->{'REST.class'}||$self->getEnv->{'PATH_INFO'}||'';
	my $method = $self->getEnv->{REQUEST_METHOD}||'';
	my $path = $self->getEnv->{SCRIPT_NAME}||'';

"
	<div class=\"header\">
		<div class=\"title\">
			<span class=\"project\">$project:</span>
			<span class=\"method\">$method</span>
			<span class=\"path\">$path</span>
		</div>
		<ul class=\"format\">
			<li><a href=\"?format=application/json\">json</a></li>
			<li><a href=\"?format=text/yaml\">yaml</a></li>
			<li><a href=\"?format=text/plain\">text</a></li>
		</ul>
	</div>
"
}

=encoding utf-8

=head1 AUTHOR

Václav Dovrtěl E<lt>vaclav.dovrtel@gmail.comE<gt>

=head1 BUGS

Please report any bugs or feature requests to github repository.

=head1 ACKNOWLEDGEMENTS

Inspired by L<https://github.com/towhans/hochschober>

=head1 REPOSITORY

L<https://github.com/vasekd/Rest-HtmlVis>

=head1 LICENSE AND COPYRIGHT

Copyright 2015 Vaclav Dovrtel.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Rest::HtmlVis::Base


lib/Rest/HtmlVis/Content.pm
package Rest::HtmlVis::Content;

use 5.006;
use strict;
use warnings FATAL => 'all';

use parent qw( Rest::HtmlVis::Key );

use Plack::Request;
use YAML::Syck;
use URI::Escape::XS qw/decodeURIComponent/;
use Encode;

$YAML::Syck::ImplicitUnicode = 1;

=head1 NAME

Rest::HtmlVis::Content - Return base block for keys links and form.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.13';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Rest::HtmlVis::Content;

    my $foo = Rest::HtmlVis::Content->new();
    ...

=head1 KEYS

=head2 link

Convert default strcuture of links. Each link should consists of:

=over 4

=item * href

URL of target.Can be absolute or relative.

=item * title

Name of the link.

=item * rel

Identifier of the link (type of the link)

=back

Example:

	link => [
		{
			href => '/api/test',
			title => 'Test resource',
			rel => 'api.test'
		}
	]

=head2 form

Define elements of formular for html.

=cut

my @defualtMethods = ('get','post','put','delete');

sub setStruct {
	my ($self, $key, $struct, $env) = @_;
	$self->{struct} = $struct;
	$self->{env} = $env;

	return 1;
}

sub getOrder {
	return 9999;
}

sub newRow {
	return 1;
}

sub head {
 return <<END;
	<script type="text/javascript">
		\$('#myTab a').click(function (e) {
		  e.preventDefault();
		  \$(this).tab('show');
		});
	function sendPut(methodType,url,jsvar){
		var selected = \$( "select[name*='enctype'] option:selected" ).val();
		var enctype = typeof selected == 'undefined'?'application/json':'application/x-www-form-urlencoded';
		\$.ajax({
			type: methodType,
			url: url,
			headers: {          
			                 Accept : 'application/json; charset=utf-8',         
			                'Content-Type': enctype + '; charset=utf-8'   
			},
			success: function(data) {
				alert('Success'); 
				window.location.href=url;
			},
			error: function(data) {
				alert(data.responseText);
			},
			data: jsvar
		});return false;
	}
	</script>
	<style>
		.pre-wrap  {
			white-space: pre-wrap;       /* css-3 */
			white-space: -moz-pre-wrap;  /* Mozilla, since 1999 */
			white-space: -pre-wrap;      /* Opera 4-6 */
			white-space: -o-pre-wrap;    /* Opera 7 */
			word-wrap: break-word;       /* Internet Explorer 5.5+ */
		}
	</style>
END
}

sub onload {
	'prettyPrint();'
}

sub html {
	my ($self) = @_;
	my $struct = $self->getStruct;
	my $env = $self->getEnv;

	### Links
	my $links = '';
	if (ref $struct eq 'HASH' && exists $struct->{link} && ref $struct->{link} eq 'ARRAY'){
		foreach my $link (@{$struct->{link}}) {
			$links .= '<li><a href="'.$link->{href}.'" rel="'.$link->{rel}.'">'.$link->{title}.'</a></li>';
		}
		delete $struct->{link};
	}

	### Remove form content
	my $formStruct; # Set undef
	$formStruct = delete $struct->{form} if (ref $struct eq 'HASH' && exists $struct->{form} && ref $struct->{form} eq 'HASH');

	### Content
	my $content = '';
	{
		local $Data::Dumper::Indent=1;
		local $Data::Dumper::Quotekeys=0;
		local $Data::Dumper::Terse=1;
		local $Data::Dumper::Sortkeys=1;

		$content = Dump($struct);
	}

	if( !$formStruct && exists $env->{'REST.class'} && $env->{'REST.class'}->can('GET_FORM') ){
		my $req = Plack::Request->new($env);
 		my $par = $req->parameters;
		$formStruct = $env->{'REST.class'}->GET_FORM($env, $content, $par);
	}

	### Form
	my $form = {};
	my $extForm = {};
	if ($formStruct){
		$form = _formToHtml($env,$formStruct,@defualtMethods);
	}
	if($formStruct){
		$extForm = _extFormToHtml($env,$formStruct,(keys %{$formStruct}));
	}
	my $firstActive = 'get';
	foreach my $method (@defualtMethods){
		if ($form->{$method}){
			$firstActive=$method;
			last;
		}
	}

	if( exists $env->{'REST.class'} && $env->{'REST.class'}->can('GET') and !exists $form->{get}){
		$form->{get} = _formToHtml($env, {get => {}}, ('get') )->{get};
	}

	my $ret = "
		<div class=\"col-lg-3\">";

			if( keys %{$extForm} ){
				$ret .= "			
							<!-- Nav tabs -->
							<ul id=\"myTabExt\" class=\"nav nav-tabs nav-justified\" role=\"tablist\">";
					my @extkey = sort(keys %{$extForm});
					my $firstActiveExt = $extkey[0];
					foreach my $method (@extkey){
						$ret .="				<li role=\"presentation\"" . ( $firstActiveExt eq $method?" class=\"active\"":($extForm->{$method}?'':" class=\"disabled\"")) . "><a role=\"tab\"". ($extForm->{$method}?" href=\"#$method\" data-toggle=\"tab\"":'') . ">" . uc($method) . "</a></li>";
					}
					$ret .= "			<!-- Tab panes -->
							<div class=\"tab-content\" id=\"myTabContent\">";
					foreach my $method (@extkey){
						my $extForPar = $extForm->{$method}{params};
						$ret .= "				<div role=\"tabpanel\" class=\"tab-pane fade" . ( $firstActiveExt eq $method?"  in active":'') . "\" id=\"$method\">
									<form class=\"method-form\" ". ($extForPar->{method} =~ /^(put|delete)$/i?"onSubmit=\""._getAjaxCall($self,  uc($extForPar->{method}),$extForPar):"method=\"" . uc($extForPar->{method})) . "\"" . (exists $extForPar->{url}?'action="'.$extForPar->{url}:'') . "\">".($extForm->{$method}{html}||'<div class="text-center"> Not allowed </div>')."
									</form>
								</div>";
					}
				  $ret .= "</div><hr />";
			}
	my $header = $self->getHeader;
	my $duration = sprintf ( "%.3f",$header->{'X-Runtime'} *1000);

  $ret .= " 			<ul class=\"links\">
				$links
			</ul>
		</div>
		<div class=\"col-lg-6\">
			<pre class=\"prettyprint lang-yaml pre-wrap\">
$content
			</pre>
			<div class=\"row align-right text-muted duration\"> <small>Duration: $duration ms </small></div>  
		</div>
		<div class=\"col-lg-3\" role=\"tabpanel\">

			<!-- Nav tabs -->
			<ul id=\"myTab\" class=\"nav nav-tabs nav-justified\" role=\"tablist\">";

	foreach my $method (@defualtMethods){
		$ret .="				<li role=\"presentation\"" . ( $firstActive eq $method?" class=\"active\"":(exists $form->{$method}?'':" class=\"disabled\"")) . "><a role=\"tab\"". (exists $form->{$method}?" href=\"#$method\" data-toggle=\"tab\"":'') . ">" . uc($method) . "</a></li>";
	}
	$ret .= "			<!-- Tab panes -->
			<div class=\"tab-content\" id=\"myTabContent\">";
	foreach my $method (@defualtMethods){
		$ret .= "				<div role=\"tabpanel\" class=\"tab-pane fade" . ( $firstActive eq $method?"  in active":'') . "\" id=\"$method\">
					<form class=\"method-form\" ". ($method =~ /^(put|delete)$/?"onSubmit=\""._getAjaxCall($self,  uc($method)):"method=\"" . uc($method)) . "\">".($form->{$method}||'<div class="text-center"> Not allowed </div>')."
					</form>
				</div>";
	}
	$ret .= "			</div></div>

		</div>
";
	return $ret;
}

sub _getAjaxCall {
	my ($self, $methodType, $params) = @_;
	my $jsvar = "\$(this).serialize()";
	my $url = $self->getEnv()->{REQUEST_URI};
	if (defined $params){
		$jsvar = $params->{jsvar} if exists $params->{jsvar};
		$url = $params->{url} if exists $params->{url};
	}
	return <<END;
	sendPut('$methodType','$url',$jsvar);
END
}


my $defaultForm = {
	get => 	"<label class=\"col-lg-4 control-label\">Get as</label> 
	<select name=\"format\" class=\"form-control\">
	  <option selected=\"selected\">text/html</option>
	  <option>application/json</option>
	  <option>text\/yaml</option>
	  <option>text/plain</option>
	</select>
	<button type=\"submit\" class=\"btn btn-default\">Get</button>",
	
	post => "<label class=\"col-lg-4 control-label\">Get as</label> 
	<select name=\"format\" class=\"form-control\">
	  <option selected=\"selected\">text/html</option>
	  <option>application/json</option>
	  <option>text/yaml</option>
	  <option>text/plain</option>
	</select>
	<label class=\"col-lg-4 control-label\">Post as</label> 
	<select name=\"enctype\" class=\"form-control\">
	  <option>application/json</option>
	  <option selected=\"selected\">text/yaml</option>
	  <option>text/plain</option>
	</select>
	<button type=\"submit\" class=\"btn btn-default\">Post</button>",

	put =>  "<label class=\"col-lg-4 control-label\">Put as</label> 
	<select name=\"enctype\" class=\"form-control\">
	  <option>application/json</option>
	  <option selected=\"selected\">text/yaml</option>
	  <option>text/plain</option>
	</select>
	<button type=\"submit\" class=\"btn btn-default\">Put</button>",

	delete => "<button type=\"submit\" class=\"btn btn-default\">Delete</button>",
};


sub _decodeQuery {
	my $str = shift || return {};
	my %ret;
	map
		{
			my ($a,$b) = split '=', $_, 2;
			unless ($a =~ /^(?:format|enctype)$/i) {
				$b =~ s/\+/ /g;
				$b = Encode::decode_utf8(decodeURIComponent($b));
				if (exists $ret{$a}) {
					$ret{$a} .= "|$b";
				} else {
					$ret{$a} = $b;
				}
			}
		}
		split '&', $str;
	return {%ret}
}

sub _paramsToHtml {
	my ($env, $param, $paramValue) = @_;
	my $type = $paramValue->{type};
	my $name = $param;
	next unless $name and $type;

	my $description = $paramValue->{description}||'';
	my $html = '';
	my $query = _decodeQuery($env->{QUERY_STRING});
	if(exists $query->{$name}){
		$paramValue->{default} = $query->{$name} =~ /\|/?[split '\|',$query->{$name}]:$query->{$name};
	}
	if ($type eq 'text'){
		my $default = ($paramValue->{default}||'');
		if (ref $default eq 'HASH' || ref $default eq 'ARRAY'){
			$default = YAML::Syck::Dump($default);
		}
		$html .= '<div class="form-group">';
		$html .= '<label>'.$description.'</label>' if ($description);
		$html .= '<input type="text" name="'.$name.'" class="form-control" value="'.$default.'"></input>';
		$html .= '</div>';
	}elsif ($type eq 'textarea'){
		my $rows = ($paramValue->{rows}||20);
		my $cols = ($paramValue->{cols}||3);
		my $default = ($paramValue->{default}||'');
		if (ref $default eq 'HASH' || ref $default eq 'ARRAY'){
			$default = YAML::Syck::Dump($default);
		}
		$html .= '<div class="form-group">';
		$html .= '<label>'.$description.'</label>' if ($description);
		$html .= '<textarea class="form-control" name="'.$name.'" rows="'.$rows.'" cols="'.$cols.'">'.$default.'</textarea>';
		$html .= '</div>';
	}elsif ($type eq 'checkbox'){
		$html .= '<div class="form-group">';
		$html .= "<label >".$description.'</label>' if ($description);
			$paramValue->{options} = [$paramValue->{options}] if ( ref $paramValue->{options} ne "ARRAY");
			foreach my $v (@{$paramValue->{options}}){
				my $optionName = ''; my $value = '';
				if (ref $v eq 'ARRAY'){
					($optionName, $value) = @$v;
				}else{
					$optionName = $v; $value = $v;
				}
				my $checked='';
				if(exists $paramValue->{default}){
					if(ref $paramValue->{default} eq 'ARRAY'){
						foreach my $d (@{$paramValue->{default}}){
							$checked = 'checked="checked"'if ($d eq $value);
						}
					}else{
						$checked = 'checked="checked"'if ($paramValue->{default} eq $value);
					}
				}
				$html .= "<div class='checkbox'><label><input type='checkbox' value='$value' name='$name' $checked />&nbsp;$optionName</label></div>";
			}
			$html .= '</div>';
	}elsif ($type eq 'radio'){
		$html .= '<div class="form-group">';
		$html .= "<label>".$description.'</label>' if ($description);
			foreach my $v (@{$paramValue->{options}}){
				my $optionName = ''; my $value = '';
				if (ref $v eq 'ARRAY'){
					($optionName, $value) = @$v;
				}else{
					$optionName = $v; $value = $v;
				}
				my $checked='';
				if(exists $paramValue->{default}){
					if(ref $paramValue->{default} eq 'ARRAY'){
						foreach my $d (@{$paramValue->{default}}){
							$checked = 'checked="checked"'if ($d eq $value);
						}
					}else{
						$checked = 'checked="checked"'if ($paramValue->{default} eq $value);
					}
				}
				$html .= "<div class='radio'><label><input type='radio' value='$value' name='$name' $checked />$optionName</label></div>";
			}
			$html .= '</div>';
	}elsif ($type eq 'select'){
		$html .= '<div class="form-group">';
		$html .= '<label>'.$description.'</label>' if ($description);
		$html .= '<select class="form-control" name="'.$name.'">';
		foreach my $v (@{$paramValue->{options}}){
			my $name = ''; my $id = '';
			if (ref $v eq 'ARRAY'){
				($id, $name) = @$v;
			}else{
				$name = $v; $id = $v;
			}
			my $default = (defined $paramValue->{default} && $id eq $paramValue->{default}) ? 'selected="selected"' : '';
			$html .= '<option id="'.$id.'" '.$default.'>'.$name.'</option>';
		}
		$html .= '</select>';
		$html .= '</div>';
	}elsif ($type eq 'hidden'){
		$html .= "  <input type='hidden' name='$param' value='$paramValue->{default}' />";
		
	}
	return $html;
}

sub _extFormToHtml {
	my ($env, $struct, @methods) = @_;
	my $form = {};
	foreach my $defaultMethod (@methods) {
		my $method = delete $struct->{$defaultMethod};
		if (defined  $method and exists $method->{params} && ref $method->{params} eq 'HASH'){
			my $html = '';
			delete $method->{params}{format};
			#delete $method->{params}{ENCTYPE};
			foreach my $param (sort {($method->{params}{$a}{description}||$a) cmp ($method->{params}{$b}{description}||$b) } keys %{$method->{params}}) {
				$html .= _paramsToHtml($env, $param, $method->{params}{$param});
			}
			$form->{$defaultMethod}{html} .= $html;
		}
		if(defined  $method and  exists $method->{default}){
			my $html = '';
			$html .= '<textarea class="form-control" name="DATA" rows="20" cols="3">'.$method->{default}.'</textarea>';
			$form->{$defaultMethod}{html} .= $html;
		}
		delete $method->{params};
		$form->{$defaultMethod}{params} = $method;
		$form->{$defaultMethod}{html} .= $defaultForm->{lc ($method->{method}) } if exists $defaultForm->{ lc($method->{method}) };
	}

	return $form;
}

sub _formToHtml {
	my ($env,$struct, @methods) = @_;
	my $form = {};
	foreach my $defaultMethod (@methods) {
		my $method = delete $struct->{$defaultMethod};
		if (defined  $method and exists $method->{params} && ref $method->{params} eq 'HASH'){
			my $html = '';
			foreach my $param (sort {($method->{params}{$a}{description}||$a) cmp ($method->{params}{$b}{description}||$b) } keys %{$method->{params}}) {
				$html .= _paramsToHtml($env, $param, $method->{params}{$param});
			}
			$form->{$defaultMethod} .= $html;
		}
		if(defined  $method and  exists $method->{default}){
			my $html = '';
			$html .= '<textarea class="form-control" name="DATA" rows="20" cols="3">'.$method->{default}.'</textarea>';
			$form->{$defaultMethod} .= $html;
		}
		
		$form->{$defaultMethod} .= $defaultForm->{$defaultMethod} if defined $method;
	}

	return $form;
}

=encoding utf-8

=head1 AUTHOR

Václav Dovrtěl E<lt>vaclav.dovrtel@gmail.comE<gt>

=head1 BUGS

Please report any bugs or feature requests to github repository.

=head1 ACKNOWLEDGEMENTS

Inspired by L<https://github.com/towhans/hochschober>

=head1 REPOSITORY

L<https://github.com/vasekd/Rest-HtmlVis>

=head1 LICENSE AND COPYRIGHT

Copyright 2015 Vaclav Dovrtel.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Rest::HtmlVis::Content
